### PR TITLE
environementd: Add Prometheus metrics to our HTTP endpoints

### PR DESCRIPTION
**Pull request by ParkMyCar**
### Motivation

Makes progress towards: #20212

This PR adds Prometheus metrics to our externally used HTTP endpoints, including the new `api/webhook/...` endpoint.

Ideally we don't have to write our own `tower::Layer`, but there doesn't seem to be any open source `Layer` that I would feel confident using. There is `axum-prometheus` but it doesn't seem to be too widely used, and this issue https://github.com/tower-rs/tower-http/issues/57 which hasn't received much traffic.

So this adds our own `PrometheusLayer` that tracks:
* Total number of requests, per-path, with status code
* Current number of inflight/concurrent requests, per-path
* Histogram of response time for requests, per-path

### Tips for reviewer

 * The diff is much smaller if viewed with whitespace hidden.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - N/a

### CODE REVIEW

File: src/environmentd/src/http.rs [hunk 1]
```diff
@@ -325,53 +332,56 @@ impl InternalHttpServer {
...
-            .route(
-                "/api/opentelemetry/config",
-                routing::put({
-                    move |_: axum::Json<DynamicFilterTarget>| async {
-                        (
-                            StatusCode::BAD_REQUEST,
-                            "This endpoint has been replaced. \
+        let router = base_router(BaseRouterConfig {
+            profiling: true,
+            metrics: None,
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/environmentd/src/http.rs]: Including it for the internal http server works for me, I added a "subsystem" to the prometheus metric to distinguish between the two. What do you think about that?

File: src/environmentd/src/http/metrics.rs [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    pub request_duration: HistogramVec,
+}
+
+impl Metrics {
+    pub(crate) fn register_into(registry: &MetricsRegistry) -> Self {
+        Self {
+            requests: registry.register(metric!(
+                name: "mz_http_requests",
+                help: "Total number of http requests received since process start.",
+                var_labels: ["path", "status"],
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/environmentd/src/http/metrics.rs:62]: Blargh you're so right, great catch!

I looked through the `axum` docs and code, nothing seems to expose the route of a request to a `Layer`, so we can't determine our endpoint that way. There were three alternatives I could think of:

1. Maintain a set of routes, match the URI of each request against a route to determine the "prometheus ready" path.
2. Log only the first ~8 characters of the path from each request, this doesn't work for the catch all route of `/` though.
3. Individually instrument each route, with a predefined name.

I went with option 3, which is more cumbersome than the first implementation where we add a layer to the entire Router, but I think is the best alternative that avoid adding arbitrary user data to the prometheus tags.

File: src/environmentd/src/http/metrics.rs [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+            )),
+            requests_active: registry.register(metric!(
+                name: "mz_http_requests_active",
+                help: "Number of currently active/open http requests.",
+                var_labels: ["path"],
+            )),
+            request_duration: registry.register(metric!(
+                name: "mz_http_request_duration",
+                help: "How long it takes for a request to complete.",
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/environmentd/src/http/metrics.rs]: Done!

File: src/environmentd/src/http/metrics.rs [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+    pub requests_active: IntGaugeVec,
+    /// How long it takes for a request to complete.
+    pub request_duration: HistogramVec,
+}
+
+impl Metrics {
+    pub(crate) fn register_into(registry: &MetricsRegistry) -> Self {
+        Self {
+            requests: registry.register(metric!(
+                name: "mz_http_requests",
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/environmentd/src/http/metrics.rs]: Done!

File: src/environmentd/src/http/metrics.rs [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+                // Record the duration of this request.
+                if let Some(timer) = this.timer.take() {
+                    timer.observe_duration();
+                }
+
+                // We've completed this request, so decrement the count.
+                this.metrics
+                    .requests_active
+                    .with_label_values(&[this.path])
+                    .dec();
```
Comment (guswynn) [inline src/environmentd/src/http/metrics.rs:222]: you'll also technically need to `dec` this in a `Drop` impl, if we weren't polled to completion

Comment (ParkMyCar) [reply src/environmentd/src/http/metrics.rs:196]: Good call! I added a Drop impl and a test to assert the right behavior

**Review by mjibson — APPROVED** [review]

**Pull request merged**